### PR TITLE
Account for potential target namespace when finding cmake module...

### PR DESCRIPTION
**Pull request by jacobperron**
for tinyxml2.

Resolves ros2/build_cop#156

### CODE REVIEW

**Review by wjwwood — APPROVED** [review]
Makes sense to me.

**Review by sloretz — DISMISSED** [review]

**Comment by jacobperron** [comment]
Looks like the namespace is not required when linking.

**Comment by jacobperron** [comment]
Test on CI machine with latest tinyxml2 release: [Build Status](https://ci.ros2.org/view/All/job/test_ci_osx/102/)

Requires [ros2/Fast-RTPS (find_tinyxml2_update)](https://github.com/ros2/Fast-RTPS/tree/find_tinyxml2_update).

**Comment by jacobperron** [comment]
Regular CI (includes MacOS with older tinyxml2 release to test backwards compatibility):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5867/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2422/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4854/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5757/)

**Comment by jacobperron** [comment]
Take 2:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5869/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2424/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4856/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5760/)

**Comment by jacobperron** [comment]
Take 3:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5870/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2425/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4857/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5761/)

**Comment by jacobperron** [comment]
The test CI job made it past the tinyxml related errors and is stuck on an unrelated error.

---

Rebuild Windows: [Build Status](https://ci.ros2.org/job/ci_windows/5762/) :crossed_fingers:

**Review by Karsten1987 — APPROVED** [review]
I contributed to this patch, but +1 because it also solves the problems locally on my osx.

File: cmake/Modules/FindTinyXML2.cmake [hunk 1]
```diff
@@ -6,10 +6,12 @@
 find_package(TinyXML2 CONFIG QUIET)
 if(TinyXML2_FOUND)
   message(STATUS "Found TinyXML2 via Config file: ${TinyXML2_DIR}")
-  if(NOT TINYXML2_LIBRARY AND TARGET tinyxml2)
+  if(NOT TINYXML2_LIBRARY)
     # in this case, we're probably using TinyXML2 version 5.0.0 or greater
     # in which case tinyxml2 is an exported target and we should use that
-    set(TINYXML2_LIBRARY tinyxml2)
+    if(TARGET tinyxml2 OR TARGET tinyxml2::tinyxml2)
+      set(TINYXML2_LIBRARY tinyxml2)
```
Comment (sloretz) [inline cmake/Modules/FindTinyXML2.cmake:13]: I don't understand. If there is a library called `tinyxml2` then I would expect a library target called `tinyxml2`. If so, `if(TARGET tinyxml2)` would also be true, and the original code would work.

Comment (wjwwood) [reply cmake/Modules/FindTinyXML2.cmake:13]: Agreed, would it sometimes be `set(TINYXML2_LIBRARY tinyxml2::tinyxml2)`?

Comment (jacobperron) [reply cmake/Modules/FindTinyXML2.cmake:13]: I also don't understand. But removing ` OR TARGET tinyxml2::tinyxml2` results in this `set` not getting called and subsequent linker errors.

Before 43522c9, I had a separate conditional block for the namespaced target, which also works:

```cmake
if(TARGET tinyxml2::tinyxml2)
  set(TINYXML2_LIBRARY tinyxml2::tinyxml2)
endif()
```

But we had an issue with this approach when applying it to Fast-RTPS, where the downstream `rmw_fastrtps_shared_cpp` package was complaining about an unresolvable target `tinyxml2::tinyxml2`... Removing the namespace from `TINYXML2_LIBRARY` resolved it. @Karsten1987 Do you have more insight as to why?

Comment (jacobperron) [reply cmake/Modules/FindTinyXML2.cmake:13]: The issue was resolved in Fast-RTPS by find_package'ing `TinyXML2` in it's cmake config.
I've updated this PR to reflect an earlier iteration.

**Review by wjwwood — APPROVED** [review]

**Review by Karsten1987 — APPROVED** [review]
rebuilding on my machine seems to work. so +1

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5882/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2432/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4864/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5772/)

**Comment by nuclearsandwich** [comment]
```
03:57:23 -- Found fastrtps_cmake_module: 0.6.0 (/home/rosbuild/ci_scripts/ws/install/fastrtps_cmake_module/share/fastrtps_cmake_module/cmake)
03:57:23 CMake Error at /home/rosbuild/ci_scripts/ws/install/fastrtps/share/fastrtps/cmake/fastrtps-config.cmake:50 (find_package):
03:57:23   By not providing "FindTinyXML2.cmake" in CMAKE_MODULE_PATH this project has
03:57:23   asked CMake to find a package configuration file provided by "TinyXML2",
03:57:23   but CMake did not find one.
03:57:23 
03:57:23   Could not find a package configuration file provided by "TinyXML2" with any
03:57:23   of the following names:
03:57:23 
03:57:23     TinyXML2Config.cmake
03:57:23     tinyxml2-config.cmake
03:57:23 
03:57:23   Add the installation prefix of "TinyXML2" to CMAKE_PREFIX_PATH or set
03:57:23   "TinyXML2_DIR" to a directory containing one of the above files.  If
03:57:23   "TinyXML2" provides a separate development package or SDK, be sure it has
03:57:23   been installed.
03:57:23 Call Stack (most recent call first):
03:57:23   CMakeLists.txt:38 (find_package)
```

**Comment by jacobperron** [comment]
After update to eProsima/Fast-RTPS#349:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5885/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2434/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4867/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5776/)

**Comment by mjcarroll** [comment]
Looks like network issues for both of these, rebuilding.

* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4874/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5785/)

**Review by sloretz — APPROVED** [review]

**Comment by nuclearsandwich** [comment]
The VCS import failed again. When I try to import that repos file locally I get

```
=== src/ros2/Fast-RTPS (git) ===
Could not checkout ref 'find_tinyxml2_update': error: pathspec 'find_tinyxml2_update' did not match any file(s) known to git.
```

Since Fast-RTPS has merged https://github.com/eProsima/Fast-RTPS/pull/349 here is the same again with just this branch.

:apple: [Build Status](https://ci.ros2.org/job/ci_osx/4875/) 
:wind_chime: (queued) [Build Status](https://ci.ros2.org/job/ci_windows/5784/)

**Comment by jacobperron** [comment]
Aaand the updated :apple: [Build Status](https://ci.ros2.org/job/test_ci_osx/105/)

**Comment by nuclearsandwich** [comment]
[Again and again...](https://www.youtube.com/watch?v=TMy6X5cQul8) :apple: [Build Status](https://ci.ros2.org/job/test_ci_osx/106/)

This time qt is linked so qmake is on PATH

**Comment by nuclearsandwich** [comment]
The CI failures above are caused by a rash of infrastructure related troubles and a recent issue with Windows that is not introduced by this change.

Build on our current MacOS configuration is passing: [Build Status](https://ci.ros2.org/job/ci_osx/4879/)
The updated MacOS configuration where this issue was first discovered is now able to build packages dependent on TinyXML2 but has issues with other upgraded packages [Build Status](https://ci.ros2.org/view/All/job/test_ci_osx/113/)
Windows test failures are unrelated [Build Status](https://ci.ros2.org/job/ci_windows/5787/) (also failures are causing builds to run very long)

I am proposing that given the reviews received on this change and the CI coverage as available, that this PR be considered merge-able.

**Pull request merged**